src/Middleware/TrustedHostsNetworkResolver.php

```diff
@@ -0,0 +1,510 @@
+                if (!isset($ipData['ip'])) {
+                    break;
+                }
+            }
+            $ip = $ipData['ip'];
+            if (!$this->isValidHost($ip, ['any'], $ipValidator)) {
+                break;
+            }
+            $ipDataList[] = $ipData;
+            if (!$this->isValidHost($ip, $trustedHostData[self::DATA_KEY_HOSTS], $ipValidator)) {
+                break;
+            }
+        } while (count($ipList) > 0);
+
+        if ($this->attributeIps !== null) {
+            $request = $request->withAttribute($this->attributeIps, $ipDataList);
+        }
+
+        $uri = $request->getUri();
+        // find HTTP host
+        foreach ($trustedHostData[self::DATA_KEY_HOST_HEADERS] as $hostHeader) {
+            if (!$request->hasHeader($hostHeader)) {
+                continue;
+            }
+            if ($hostHeader === $ipHeader && $ipListType === self::IP_HEADER_TYPE_RFC7239 && isset($ipData['httpHost'])) {
+                $uri = $uri->withHost($ipData['httpHost']);
+                break;
+            }
+            $host = $request->getHeaderLine($hostHeader);
+            if (filter_var($host, FILTER_VALIDATE_DOMAIN) !== false) {
+                $uri = $uri->withHost($host);
+                break;
+            }
+        }
+
+        // find protocol
+        foreach ($trustedHostData[self::DATA_KEY_PROTOCOL_HEADERS] as $protocolHeader => $protocols) {
+            if (!$request->hasHeader($protocolHeader)) {
+                continue;
+            }
+            if ($protocolHeader === $ipHeader && $ipListType === self::IP_HEADER_TYPE_RFC7239 && isset($ipData['protocol'])) {
+                $uri = $uri->withScheme($ipData['protocol']);
+                break;
+            }
+            $protocolHeaderValue = $request->getHeaderLine($protocolHeader);
+            foreach ($protocols as $protocol => $acceptedValues) {
+                if (\in_array($protocolHeaderValue, $acceptedValues)) {
+                    $uri = $uri->withScheme($protocol);
+                    break 2;
+                }
+            }
+        }
+        $urlParts = $this->getUrl($request, $trustedHostData[self::DATA_KEY_URL_HEADERS]);
+        if ($urlParts !== null) {
+            [$path, $query] = $urlParts;
+            $uri = $uri->withPath($path);
+            if ($query !== null) {
+                $uri = $uri->withQuery($query);
+            }
+        }
+        return $handler->handle($request->withUri($uri)->withAttribute('requestClientIp', $ipData['ip']));
+    }
+
+    /**
+     * Validate host by range
+     *
+     * This method can be extendable by overwriting eg. with reverse DNS verification.
+     */
+    protected function isValidHost(string $host, array $ranges, Ip $validator): bool
+    {
+        return $validator->ranges($ranges)->validate($host)->isValid();
+    }
+
+    /**
+     * Reverse obfuscating host data
```

Comment:
It's not clear what "reverse obfuscation" stands for. Could you elaborate?

Reply:
more php doc added